### PR TITLE
feat(Dunning Enhancements): Dunning type, resolution

### PR DESCRIPTION
**Pull request by KanchanChauhan**
Enhancements Done:
1. Added Company field in case there is a multi-company setup.
2. Filtered Sales Invoice based on Company and if they are unpaid.

Screen Shot 2019-12-04 at 12 36 20 PM

3. There can be different types of dunning like First Overdue Notice, Second Overdue Notice and based on that can be the different interest rates and dunning fees. Added Dunning Type doctype. It will have a name, interest rate, and dunning fee.
Screen Shot 2019-12-04 at 12 40 40 PM

4. Add Dunning Type field to Dunning and pull Dunning Fee and Rate of Interest from Dunning Type
p

5. Added field Resolution to check if the particular Dunning is resolved or not
Screen Shot 2019-12-04 at 12 43 19 PM

6. List view changes with an indicator to see if Dunning is resolved or not
Screen Shot 2019-12-04 at 12 43 28 PM

I am a Software Developer with ESO Electronic, Germany. Please let me know if you have any questions regarding this PR or need any clarification.

### CODE REVIEW

**Comment by KanchanChauhan** [comment]
Will be removing some custom stuff soon.

File: dunning/dunning/doctype/dunning/dunning.js [hunk 1]
```diff
@@ -2,12 +2,33 @@
...
 	refresh: function (frm) {
+		frm.set_df_property("company", "read_only", frm.doc.__islocal ? 0 : 1);
+		frm.toggle_display("naming_series", false);
 		if (frm.is_new()) {
 			frm.trigger("calculate_overdue_days");
 			frm.set_value("posting_date", frappe.datetime.nowdate());
 		}
 	},
+  	company: function(frm) {
+		if (frm.doc.company == 'SC ESO Electronic S.R.L') {
```
Comment (barredterra) [inline dunning/dunning/doctype/dunning/dunning.js]: This shouldn't be hardcoded ;) I think naming series should be a Select field, as in the other accounting doctypes

File: dunning/dunning/doctype/dunning_type/dunning_type.json [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+   "fieldtype": "Float",
+   "in_list_view": 1,
+   "label": "Interest Rate",
+   "reqd": 1
+  },
+  {
+   "fieldname": "dunning_type",
+   "fieldtype": "Data",
+   "in_list_view": 1,
+   "label": "Name",
```
Comment (barredterra) [inline dunning/dunning/doctype/dunning_type/dunning_type.json]: Please use "Title" or "Dunning Type" as the label. "Name" usually refers to the unique ID.

**Review by barredterra — CHANGES_REQUESTED** [review]

File: dunning/dunning/doctype/dunning/dunning.js (hunk 1 above)
Comment (KanchanChauhan) [reply dunning/dunning/doctype/dunning/dunning.js]: Removed the custom code. :)

File: dunning/dunning/doctype/dunning_type/dunning_type.json (hunk 2 above)
Comment (KanchanChauhan) [reply dunning/dunning/doctype/dunning_type/dunning_type.json]: Named it as Dunning Type

**Comment by KanchanChauhan** [comment]
@barredterra Autoname changed to Naming series like in other doctypes. Default is PR-MM-YY (Payment Reminder - Month - Year - series)
Screen Shot 2019-12-05 at 9 27 23 AM
Also added a letter head field at the end, when you have letterhead field, print formats automatically takes the header and the footer and we don't have to get them in jinja. Check out the print format in this PR. It is just basic but can be modified.
Screen Shot 2019-12-05 at 9 19 59 AM
And added currency to Print Settings which will be pulled from Sales Invoice, makes it easy to get currency symbol with amounts. Check out the print format in this PR.
Screen Shot 2019-12-05 at 10 15 43 AM

File: dunning/dunning/print_format/dunning/dunning.json [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
...\n            <td style=\\\"text-align: right;\\\">\\n                {{frappe.db.get_value(\\\"Currency\\\",doc.currency, \\\"symbol\\\")}} {{doc.get_formatted(\\\"interest_amount\\\")}}\\n            </td>\\n        </tr>\\n        {% endif %}\\n        {%if doc.dunning_fee > 0%}\\n        <tr>\\n            <td>\\n                {{_(\\\"Dunning Fee\\\")}}\\n             </td>\\n            <td style=\\\"text-align: right;\\\">\\n                {{frappe.db.get_value(\\\"Currency\\\",doc.currency, \\\"symbol\\\")}} {{doc.get_formatted(\\\"dunning_fee\\\")}}\\n            </td>\\n        </tr>\\n        {% endif %}\\n    </tbody>\\n</table>    \"}, {\"fieldname\": \"_custom_html\", \"print_hide\": 0, \"label\": \"Custom HTML\", \"fieldtype\": \"HTML\", \"options\": \"\\n<div class=\\\"col-xs-6\\\">\\n</div>\\n<div class=\\\"col-xs-6\\\">\\n<div class=\\\"row total\\\">\\n\\t\\t<div class=\\\"col-xs-5\\\">\\n\\t\\t\\t<b>{{_(\\\"Total\\\")}}</b></div>\\n\\t\\t<div class=\\\"col-xs-7 text-right\\\" style=\\\"padding-right:0px;\\\">\\n\\t\\t\\t<b>{{frappe.db.get_value(\\\"Currency\\\",doc.currency, \\\"symbol\\\")}} {{doc.get_formatted(\\\"sum\\\")}}</b>\\n\\t\\t</div>\\n</div>\"}]",
```
Comment (barredterra) [inline dunning/dunning/print_format/dunning/dunning.json:11]: The print format still contains custom strings ;)

File: dunning/dunning/doctype/dunning/dunning_list.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
+frappe.listview_settings['Dunning'] = {
+	get_indicator: function (doc) {
+		if (doc.resolution_status === "Resolved") {
+			return [__("Resolved"), "green", "status,=,Resolved"];
```
Comment (barredterra) [inline dunning/dunning/doctype/dunning/dunning_list.js:4]: How does the `resolution_status` change to "Resolved"?

A Dunning for a paid Sales Invoice is still marked as "Unresolved".

**Review by barredterra — CHANGES_REQUESTED** [review]

File: dunning/dunning/print_format/dunning/dunning.json (hunk 3 above)
Comment (barredterra) [reply dunning/dunning/print_format/dunning/dunning.json:11]: The sender address could be made dynamic like this (incomplete example!):
```jinja
{% set cmp_addr_name = frappe.db.get_value("Address", {"is_your_company_address": 1}, "name") %}
{% set cmp_addr = frappe.get_doc("Address", cmp_addr_name) %}

<p>{{ doc.company }} &#183; {{ cmp_addr.address_line1 }} &#183;
           {% if cmp_addr.address_line2 %}{{ cmp_addr.address_line2 }} <br>{% endif %}
           {{ cmp_addr.pincode }} {{ cmp_addr.city }}
           {% if cmp_addr.country != pay_addr.country -%}
                &#183; {{ cmp_addr.country | upper }}
           {%- endif %}
</p>
```

**Comment by barredterra** [comment]
@KanchanChauhan are you still working on this?

File: dunning/dunning/doctype/dunning/dunning_list.js (hunk 4 above)
Comment (KanchanChauhan) [reply dunning/dunning/doctype/dunning/dunning_list.js:4]: This will be changed manually by the user.

**Comment by KanchanChauhan** [comment]
@barredterra I have some changes on my local, I will push them tomorrow and then we can move forward with this PR.

**Comment by barredterra** [comment]
@KanchanChauhan are you still working on this?

- This needs a patch that fills new mandatory fields in existing Dunnings (company). Otherwise the status cannot be changed for existing Dunnings because validation fails.
- Changing the status doesn't work on a fresh installation.
- After making status editable, status in list view doesn't get updated.

**Comment by barredterra** [comment]
@scdanieli continued the work on this. Closing this PR in favor of #2.

**Pull request closed without merging**